[DEV-17234] Update marketplace menu item

michalkleiner opened the pull request:
### Description:

Ref. DEV-17234

@sgiehl can you check the approach here? It's based on Marc's investigation into how this might be possible to achieve and it seems to be working ok.

The approach is to allow any category to pass a vue component widget string that gets translated into a component when the menu is constructed.

The click event is reused from the category so the target widget loads in the righthand side area as expected.

I've also added the outline link to the UI demo view.

Remaining to do:
- update icon font adding the shopping basket icon
- use the basket icon in the menu button
- update UI tests

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

michalkleiner committed:
Fix CS

michalkleiner committed:
Fix typo in UI demo example

michalkleiner committed:
Add styling and UI demo example for outline button (with reverse bg on hover)

michalkleiner committed:
Remove premium features widget

michalkleiner committed:
Allow category to have a custom widget for the reporting menu

michalkleiner committed:
Implement marketplace rich menu button

michalkleiner committed:
Build vue files

michalkleiner committed:
Remove surplus category config

michalkleiner commented on plugins/CoreHome/vue/src/ReportingMenu/ReportingMenu.vue:
```diff
@@ -12,13 +12,18 @@
         role="menuitem"
         v-for="category in menu"
-        :class="{ 'active': category.id === activeCategory }"
+        :class="`menuTab ${category.id === activeCategory ? 'active' : null} ${category.id}`"
```
Outputting the category ID may need some sanitisation in case it's not CSS class name compatible.

sgiehl replied:
What is the category id as class name actually good for? You could also use a custom html attribute for that. That would allow any value and could be used in CSS as well.

michalkleiner replied:
Yeah, it's to apply a specific styling. We can use a custom data attribute, or convert the ID to a CSS-able string. The result and use is the same in the end. I guess a data attribute can stomach more content than a CSS class name.

michalkleiner replied:
replaced with a data attribute

michalkleiner commented on plugins/CoreHome/vue/src/ReportingMenu/ReportingMenu.vue:
```diff
@@ -93,14 +98,20 @@
     >
       <li
-        class="no-padding"
+        :class="`no-padding ${category.id}`"
```
as above, may need escaping

michalkleiner replied:
replaced with a data attribute

michalkleiner commented:
@sgiehl can you please provide any feedback you may have on the approach? Is it ok to leave it like this or do we need to change it substantially (if so, how)?

michalkleiner committed:
Update unit tests with an empty category widget prop

sgiehl commented on core/Category/Category.php:
```diff
@@ -121,6 +128,17 @@ public function getIcon()
         return $this->icon;
     }
 
+    public function setWidget($widget)
+    {
+        $this->widget = $widget;
+        return $this;
+    }
+
+    public function getWidget()
+    {
+        return $this->widget;
+    }
```
Those methods could have some type hinting

michalkleiner replied:
added

sgiehl commented on plugins/Marketplace/lang/en.json:
```diff
@@ -84,6 +84,7 @@
         "PriceExclTax": "%1$s %2$s excl. tax.",
         "PriceFromPerPeriod": "From %1$s \/ %2$s",
         "Reviews": "Reviews",
+        "RichMenuIntro": "Explore and purchase essential plugins for your dashboard.",
```
Did you check if the translations of the removed categories and page are still in use somewhere else? Otherwise they could be removed in that file...

michalkleiner replied:
I've reinstated the paid plugins widget as it's used in other places (or at least the UI tests indicate that), and just removed the reporting menu category itself, so I think the translations can stay. Haven't done a thorough sweep though.

sgiehl reviewed (changes requested):
Had a rough look through the code. The approach looks fine to me that way.
Left a couple of comments for things that came to my mind when looking through the changes...

michalkleiner committed:
Update system test expected files

michalkleiner committed:
Add type hints

michalkleiner committed:
Update UI test screenshots

michalkleiner committed:
Update expected API response xml

michalkleiner committed:
Partially revert "Remove premium features widget"

This reverts commit 5f37b028cf6ad872e58065a64cbda22259d4d19a.
We are not reverting the removal of the subcategory.

michalkleiner committed:
built vue files

michalkleiner committed:
Restore Paid Plugins widget in system test expected files

Partially reverts commit c351ec94e8bb3883e7b753aefae28ce3e7386862.

michalkleiner committed:
Update expected test files

michalkleiner committed:
Use data attribute instead of CSS class

michalkleiner committed:
Add tabindex

michalkleiner committed:
Build vue files

michalkleiner committed:
built vue files

michalkleiner committed:
Let category load when there's exactly one subcategory or multiple subcategories and an override widget

michalkleiner committed:
Add support for keyboard submit

michalkleiner committed:
Build vue files

michalkleiner committed:
Update quick access search UI test screenshot

PM is ok with the menu items being removed from the quick access menu. Keyboard access to the new rich menu button works.

michalkleiner committed:
Update built vue files

michalkleiner commented on tests/UI/expected-screenshots/QuickAccess_search_1.png:
Confirmed with PM that the quick access items for Marketplace are OK to be removed as long as the new rich button in menu nav is accessible using a keyboard.

michalkleiner commented on plugins/CoreHome/vue/src/ReportingMenu/ReportingMenu.vue:
```diff
@@ -283,10 +307,13 @@ export default defineComponent({
+
+      // one subcategory or a widget and some subcategories to allow to load the category
+      const { subcategories } = category as SubcategoryContainer;
+      const categoryCanLoad = (subcategories && subcategories.length === 1)
+        || (category.widget && subcategories && subcategories.length);
```
I know I could extract the `subcategories` condition to the front though it seems better for legibility to keep them bunched like this.

caddoo replied:
I'm wondering why in the second check ` || (category.widget && subcategories && subcategories.length);` 

that `subcategories && subcategories.length` is needed? Isn't there only being a widget enough?

michalkleiner replied:
No, the widget itself doesn't do anything, it actually triggers the action of the first nested menu item. Nested menu items are subcategories, so we need at least one to be there (but can be multiple).

So the logic is:
- no widget override, exactly one subcategory -> we load the view for the first (only) item
- no widget override, more subcategories -> we open the nested menu and no change of view
- widget override, one or more subcategories -> we load the view for the first item

michalkleiner replied:
There's surely more ways how the widget override can be integrated within the reporting menu, e.g. we can work with the subcategories within the override component, or it can have the link hardcoded etc. At the moment it takes the marketplace menu item and uses its link, so if that changes coming from the backend, the override will still work.

caddoo replied:
Thanks for clarifying

caddoo commented on plugins/API/tests/Unit/WidgetMetadataTest.php:
```diff
@@ -48,6 +48,7 @@ public function test_buildWidgetMetadata_ShouldGenerateMetadata()
                 'order' => 99,
                 'icon' => '',
                 'help' => '',
+                'widget' => null,
```
Probably a good idea to test this with an actual value at least once.

As you are missing a test for this logic:
`'widget' => $category->getWidget() ?: null,`

michalkleiner replied:
Yep, I can probably update one of the system tests where it returns the API response in XML format for comparison.

Otherwise the UI tests for the menu already cover the visual component override depending on this working.

caddoo commented on plugins/CoreHome/vue/src/ReportingMenu/ReportingMenu.vue:
```diff
@@ -180,7 +195,16 @@ export default defineComponent({
+      const categories = ReportingMenuStoreInstance.menu.value;
+
+      categories.forEach((category) => {
+        if (category.widget && category.widget.indexOf('.') > 0) {
+          const [widgetPlugin, widgetComponent] = category.widget.split('.');
```
```suggestion
        const widgetParts = category.widget.split('.');
        if (category.widget && widgetParts.length ===  2) {
          const [widgetPlugin, widgetComponent] = widgetParts;
```

Just to prevent bugs/issues later when there are `category.widget` values of things like `a.b.c`

michalkleiner replied:
I can add a limit to the split function so it always produces two parts from the first two, or we use splat to catch all the remaining parts into the component.. or I can add a check to the `setWidget` call that would throw on invalid format inputs... basically I don't think it should be dealt with here in this way, but happy to discuss. It's basically a matter "if someone provides incompatible format widget notation, it won't work for them" and them is mostly us.

caddoo replied:
Maybe in the backend is better `setWidget` can be a run time error at least and stop any headaches going forward. A bit of defensive programming.
Otherwise it could be a value object with two parts `plugin` and `component` and a __toString() method that concats them in the correct format.

caddoo commented:
I'm still a bit hesitant to approve this, but I did test it locally and it's all good. Maybe @sgiehl you wouldn't mind taking a look as you looked at the approach earlier in the PR

sgiehl commented on plugins/CoreHome/vue/src/ReportingMenu/ReportingMenu.vue:
```diff
@@ -93,14 +100,21 @@
     <ul
       id="mobile-left-menu"
-      class="sidenav hide-on-large-only"
+      class="sidenav sidenav--reporting-menu-mobile hide-on-large-only"
```
Any reason for adding a new class name here instead of using the existing ones or maybe the id? Just wondering why the already available attributes shouldn't be enough to select the element in CSS. If the available classes aren't unique enough you could simply use the class `reportingMenu` of the surrounding div.

Also the double-dash class name looks really ugly. I know we have a couple other places where that is used, but imho we should maybe try to get rid of them instead of adding new ones.

michalkleiner replied:
Thanks for the comment @sgiehl!

Probably didn't need the new class, that's true. There's so many different ways of targeting elements in CSS. I don't think using the ID is the right way and I thought the sidenav wasn't unique to this element, and a combination of e.g. `.sidenav.hide-on-large-only` is also not the prettiest.
The `--` double dash is a modifier syntax from the BEM methodology. Although there might be a couple of similar uses in other places, it's true it wasn't necessary here either.

We should perhaps standardise how we do CSS and front-end naming in general and then we can have a common and shared approach to new additions, otherwise it'll be a mix of things.

sgiehl reviewed (commented):
Overall looks good to me and works as expected. Left a minor comment for a possible improvement. Otherwise this looks good to merge.
